### PR TITLE
Output file from query in non-interactive mode

### LINKED ISSUE #306: Feature request: Support -o to output a file

### PR DESCRIPTION
**Pull request by ellbosch**
Closes #306.

### CODE REVIEW

File: mssqlcli/mssql_cli.py [hunk 1]
```diff
@@ -187,31 +183,22 @@ def __init__(self, options):
...
+        if self.interactive_mode and self.output_file:
+            click.secho("Invalid request: -o must be accompanied with a query.", err=True, fg='red')
+            sys.exit(1)
+
     def __del__(self):
         # Shut-down sqltoolsservice
         if self.sqltoolsclient:
             self.sqltoolsclient.shutdown()
 
-    def write_to_file(self, pattern, **_):
```
Comment (ellbosch) [inline mssqlcli/mssql_cli.py]: This doesn't get used by anything, but may eventually be used for \o special query. @pensivebrian--up to you to keep it or not.

File: mssqlcli/mssql_cli.py [hunk 2]
```diff
@@ -372,16 +359,14 @@ def _output_query(self, output):
...
-            #             click.echo('\n'.join(output), file=f)
-            #             click.echo('', file=f)  # extra newline
-            #     except IOError as e:
-            #         click.secho(str(e), err=True, fg='red')
-            click.echo('\n'.join(output))
+            if self.output_file:
+                try:
+                    with open(self.output_file, 'w', encoding='utf-8') as f:
+                        click.echo('\n'.join(output), file=f)
+                except IOError as e:
```
Comment (pensivebrian) [inline mssqlcli/mssql_cli.py]: Shouldn't we bubble up a non-zero return code if we encounter an exception?

File: mssqlcli/mssql_cli.py (hunk 1 above)
Comment (pensivebrian) [reply mssqlcli/mssql_cli.py]: I'd leave it if the \o special command uses it.

File: tests/test_noninteractive_mode.py [hunk 3]
```diff
@@ -12,6 +12,15 @@ class TestNonInteractiveResults:
     """
     Tests non-interactive features.
     """
+    @staticmethod
+    @pytest.fixture()
+    def tmp_filepath():
+        """ pytest fixture which returns filepath string and removes the file after tests
+        complete. """
+        fp = _FILE_ROOT + "tmp.txt"
```
Comment (pensivebrian) [inline tests/test_noninteractive_mode.py]: There should be an API to get a unique temp file name.

File: mssqlcli/mssqlclioptionsparser.py [hunk 4]
```diff
@@ -144,6 +144,14 @@ def create_parser():
         help=u'Executes a query outputting the results to stdout and exits.'
     )
 
+    args_parser.add_argument(
+        u'-o', u'--output_file',
+        dest=u'output_file',
+        metavar=u'',
+        default=False,
+        help=u'Identifies the file that receives output from a query.'
```
Comment (pensivebrian) [inline mssqlcli/mssqlclioptionsparser.py]: Identifies = Specifies

File: mssqlcli/mssql_cli.py [hunk 5]
```diff
@@ -391,6 +376,13 @@ def run(self):
             raise ValueError("'run' must be used in interactive mode! Please set \
                              interactive_mode to True.")
 
+        # exit and return error if user enters interactive mode with -o argument enabled
+        if self.output_file:
+            click.secho("Invalid request: -o must be accompanied with a query with -Q.",
```
Comment (pensivebrian) [inline mssqlcli/mssql_cli.py]: "invalid request" maybe should be "invalid arguments" ?

File: mssqlcli/mssql_cli.py [hunk 6]
```diff
@@ -187,31 +183,22 @@ def __init__(self, options):
...
 
+        # output file is for non-interactive mode
+        self.output_file = options.output_file
+
         self.sqltoolsclient = SqlToolsClient(enable_logging=options.enable_sqltoolsservice_logging)
         self.mssqlcliclient_main = MssqlCliClient(options, self.sqltoolsclient)
 
+        # exit and return error if user enters interactive mode with -o argument enabled
+        if self.interactive_mode and self.output_file:
+            click.secho("Invalid request: -o must be accompanied with a query.", err=True, fg='red')
```
Comment (pensivebrian) [inline mssqlcli/mssql_cli.py]: "Invalid request" = "Invalid Arguments"?

File: tests/test_noninteractive_mode.py (hunk 3 above)
Comment (pensivebrian) [inline tests/test_noninteractive_mode.py]: Can we name _FILE_ROOT a little better? Maybe _BASELINE_DIR?

**Pull request merged**